kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaEvictionService.java

```diff
@@ -0,0 +1,164 @@
+    replicaMarkEvictTimer = meterRegistry.timer(REPLICA_MARK_EVICT_TIMER);
+  }
+
+  @Override
+  protected void runOneIteration() {
+    markReplicasForEviction();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        managerConfig.getReplicaEvictionServiceConfig().getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting replica eviction service");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closed replica eviction service");
+  }
+
+  /**
+   * Marks cache slots that contain an expired replica as ready for eviction. Once eviction has
+   * started the cache node will further update the state indicating it has started evicting the
+   * replica from memory/disk.
+   */
+  protected int markReplicasForEviction() {
+    Timer.Sample evictionTimer = Timer.start(meterRegistry);
+
+    Map<String, ReplicaMetadata> replicaMetadataByReplicaId =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            .collect(Collectors.toUnmodifiableMap(ReplicaMetadata::getName, Function.identity()));
+
+    long expireAfterNow = Instant.now().toEpochMilli();
+    AtomicInteger successCounter = new AtomicInteger(0);
+    List<ListenableFuture<?>> replicaEvictions =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            // get all the slots that are currently assigned, but have an expiration in the past
+            .filter(
+                cacheSlotMetadata ->
+                    cacheSlotMetadata.cacheSlotState.equals(
+                            Metadata.CacheSlotMetadata.CacheSlotState.ASSIGNED)
+                        && replicaMetadataByReplicaId.containsKey(cacheSlotMetadata.replicaId)
+                        && replicaMetadataByReplicaId.get(cacheSlotMetadata.replicaId)
+                                .expireAfterUtc
+                            < expireAfterNow)
+            .map(
+                (cacheSlotMetadata) -> {
+                  ListenableFuture<?> future =
+                      cacheSlotMetadataStore.update(
+                          new CacheSlotMetadata(
+                              cacheSlotMetadata.name,
+                              Metadata.CacheSlotMetadata.CacheSlotState.EVICT,
+                              cacheSlotMetadata.replicaId,
+                              Instant.now().toEpochMilli()));
+
+                  addCallback(
+                      future,
+                      successCountingCallback(successCounter),
+                      MoreExecutors.directExecutor());
+                  return future;
+                })
+            .collect(Collectors.toUnmodifiableList());
+
+    ListenableFuture<?> futureList = Futures.successfulAsList(replicaEvictions);
```

Comment:
Since this code is repetitive across services, let's refactor it into a utility function.

Reply:
As discussed, I agree that we should come up with a common util. Let's address this as a separate PR once the pending ones are merged as a single pass. See https://github.com/slackhq/kaldb/pull/176#discussion_r779696717 for discussion about where I'm thinking we should go with this.